src/plugins/analysis/plugin.py

```diff
@@ -0,0 +1,393 @@
+                value,
+            )
+            if not is_semver:
+                raise ValueError('Version must use semantic versioning')
+
+    def __init__(self, metadata: MetaData):
+        self.metadata: PluginV1.MetaData = metadata
+
+    # The type MetaData.Schema
+    Schema = typing.TypeVar('T')
+
+    @abc.abstractmethod
+    def summarize(self, result: Schema) -> list[str]:
+        """Returns a list of  strings that summarizes the analysis.
+        Will only be called if analyze did not return None.
+
+        A summary most often is either just a single sentence or a list of words.
+        Some examples of summarys are:
+
+            * From the software_components plugin: ``["BusyBox 1.29.3", "Linux Kernel 4.9.250", "SQLite 3.8.11.1"]``
+            * ``["Nothing found"]``
+            * ``["SSLCertificate"]``
+
+        :param result: The analysis as returned by :py:func:`analyze`
+        """
+
+    @abc.abstractmethod
+    def analyze(self, file_handle: io.FileIO, virtual_file_path: dict, analyses: dict[str, dict]) -> Schema:
+        """Analyze a file.
+
+        :param file_handle: :py:class:`io.FileIO` instance of the file to be analyzed
+        :param virtual_file_path: The virtual file paths, see :py:class:`~objects.file.FileObject`
+        :param analyses: A dictionary of dependent analysis
+        """
+
+    @typing.final
+    def get_analysis(self, file_handle: io.FileIO, virtual_file_path: dict, analyses: dict[str, dict]) -> dict:
+        start_time = time.time()
+        result = self.analyze(file_handle, virtual_file_path, analyses)
+
+        if result is not None:
+            summary = self.summarize(result)
+            tags = self.get_tags(result)
+        else:
+            summary = ['Nothing found.']
+            tags = []
+
+        # The dictionary as defined in the docs for FileObject.analyses_tags
+        # Misses the root uid, which must be added by the scheduler
+        tag_dict = {}
+        for tag in tags:
+            name = tag.pop('name')
+            tag_dict.update({name: tag})
+
+        return dict(
+            analysis_date=start_time,
+            plugin_version=self.metadata.version,
+            system_version=self.metadata.version,
+            summary=summary,
+            # TODO in what case dont we want to overwrite the old tags?
+            # compare to add_analysis_tag
+            tags=tag_dict,
+            result=result,
+        )
+
+    def get_tags(self, result: Schema) -> list[dict]:
+        """Returns a list of tags to be added to the firmware.
+        The strucutre of the dict is defined in the docs for :py:attr:`objects.FileObject.analysis_tags`.
+
+        The dict contains an extra key: "name".
+        This key is the name of the tag.
+        """
+        del result
+        return []
+
+
+class PluginRunner:
```

Comment:
`PluginRunner` and `Worker` should be moved to the scheduler, shouldn't they?
They don't have anything to do with defining plugins, just with running them (which could in theory be done otherwise than `PluginRunner` does).